core/metacling/src/TCling.cxx

```diff
@@ -9376,15 +9376,28 @@ void TCling::ForgetMutexState()
 
 void TCling::ApplyToInterpreterMutex(void *delta)
 {
-   R__ASSERT(!fInitialMutex.empty() && "Inconsistent state of fInitialMutex!");
    if (gInterpreterMutex) {
       if (delta) {
-         auto typedDelta = static_cast<TVirtualRWMutex::StateDelta *>(delta);
-         std::unique_ptr<TVirtualRWMutex::StateDelta> uniqueP{typedDelta};
-         gCoreMutex->Apply(std::move(uniqueP));
+         auto typedDelta = static_cast<MutexStateAndRecurseCountDelta *>(delta);
+         std::unique_ptr<MutexStateAndRecurseCountDelta> uniqueP{typedDelta};
+         gCoreMutex->Apply(std::move(typedDelta->fDelta));
+         // Now that we have the lock, update the global
+         R__ASSERT(fInitialMutex.fRecurseCount == 0 && "Inconsistent state of fInitialMutex!  Another thread within Interpreter critical section.");
+         std::swap(fInitialMutex, typedDelta->fInitialState);
+      } else {
+         // This case happens whenever, EnableThreadSafety is first called from
+         // the interpreter function we just handled.
+         // Since thread safetely was not enabled at the time we rewinded, there was
+         // no lock taken and even-though we should be locking the rest of this
+         // interpreter handling/modifying code (since there might be threads in
+         // flight), we can't because there would be any lock guard to release the
+         // locks
+         if (fInitialMutex || fInitialMutex.fRecurseCount !=0)
+            Error("ApplyToInterpreterMutex",
```

Comment:
Could `EnableThreadSafety` inform `TCling` about it being called, such that `TCling` can do "whatever is necessary" in case it was invoked from the interpreter? If so, this review comment could end up in a new GH issue; I find the situation is handled well enough in this PR.

Reply:
Well .. 'what ever is necessary' involves going back up in the stack and update stack objects (the TLockGuard) and update them to point to the newly created mutex and also ... count them to properly invent the lock state. I suspect this can only be done by introduce a non-locking implementation of the mutex interface that does the counting ....